File: leetcode/68.text-justification.cc
class Solution {
public:
    vector<string> fullJustify(vector<string> &words, int L) {
        vector<string> res;
        // if(L==0) return words;
        size_t n=words.size();
        size_t start=0;
        while(start<n) {
            size_t ls=0;
            size_t i=start;
            for(; i<n && ls+i-start+words[i].size()<=L; i++) {
                ls+=words[i].size();
            }
            
            if(i==start+1) { // only one word
                string line = words[start];
                int endspace=L- (int) line.size(); // line needs to fill in space at the end
                line.append(endspace, ' ');
                res.push_back(line);
                start++;
            } else { // i will be index of the next word
                size_t space=(L-ls)/(i-start-1);
                size_t bonus=(L-ls)-space*(i-start-1);
                if(i>=n) { // last line
                    bonus=0;
                    space=1;
                }
                string line=words[start++];
                while(start<i) {
                    if(bonus>0){
                        line.append(space+1, ' ');
                        bonus--;
                    } else {
                        line.append(space, ' ');
                    }
                    line+=words[start++];
                }
                int endspace=L-line.size();// corner case: last line needs to fill in space at the end
                line.append(endspace, ' ');
                res.push_back(line);
                start=i;
            }
        }
        return res;
    }
};

File: leetcode/68.text-justification.java
public class Solution {
    public List<String> fullJustify(String[] words, int maxWidth) {
        int n = words.length;
        int start = 0;
        List<String> res = new ArrayList<>();
        while(start<n) {
            int end=start;
            int ls=0;
            for(;end<n && ls+end-start+words[end].length()<=maxWidth;end++) {
                ls+=words[end].length();
            }
            
            StringBuilder sb = new StringBuilder();
            if(start+1==end) {
                sb.append(words[start]);
                if(maxWidth-words[start].length()>0) {
                    appendSpace(maxWidth-words[start].length(), sb);
                }
            } else {
                int space, bonus;
                if(end>=n) {
                    space = 1;
                    bonus = 0;
                } else {
                    space = (maxWidth-ls)/(end-start-1);
                    bonus = (maxWidth-ls)-space*(end-start-1);
                }
                sb.append(words[start]);
                while(start+1<end) {
                    if(bonus>0) {
                        appendSpace(space+1, sb);
                        bonus--;
                    } else {
                        appendSpace(space, sb);
                    }
                    sb.append(words[++start]);
                }
                if(maxWidth-sb.length()>0) {
                    appendSpace(maxWidth-sb.length(), sb);
                }
             }
             res.add(sb.toString());
             start=end;
        }
        return res;
    }
    
    void appendSpace(int n, StringBuilder sb) {
        char[] ca = new char[n];
        Arrays.fill(ca, ' ');
        sb.append(ca);
    }
}
